File: algorithms/1544.make-the-string-great/solution.rb
# @param {String} s
# @return {String}
def make_good(s)
  stack = []
  s.chars.each do |c|
    if !stack.empty? && ((stack[-1].is_upper? && stack[-1].downcase == c) || (stack[-1].is_lower? && stack[-1].upcase == c))
      stack.pop
    else
      stack.push(c)
    end
  end
  stack.join
end

class String
  def is_upper?
    self >= "A" && "Z" >= self
  end

  def is_lower?
    self >= "a" && "z" >= self
  end
end

File: algorithms/1544.make-the-string-great/solution.rs
impl Solution {
    pub fn make_good(s: String) -> String {
        let mut stack: Vec<char> = vec![];
        for c in s.chars() {
            if let Some(last) = stack.last() {
                if (last.is_uppercase() && last.to_ascii_lowercase() == c)
                    || (last.is_lowercase() && last.to_ascii_uppercase() == c)
                {
                    stack.pop();
                    continue;
                }
            }
            stack.push(c);
        }
        stack.iter().collect()
    }
}
